encryption_jwt/codec_server.py

```diff
@@ -0,0 +1,119 @@
+AUTHORIZED_NAMESPACE_ACCESS_ROLES = ["read", "write", "admin"]
+
+temporal_ops_address = "saas-api.tmprl.cloud:443"
+if os.environ.get("TEMPORAL_OPS_ADDRESS"):
+    temporal_ops_address = os.environ.get("TEMPORAL_OPS_ADDRESS")
+
+
+def build_codec_server() -> web.Application:
+    # Cors handler
+    async def cors_options(req: web.Request) -> web.Response:
+        resp = web.Response()
+
+        if req.headers.get(hdrs.ORIGIN) == "http://localhost:8080":
+            logger.info("Setting CORS headers for localhost")
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_ORIGIN] = "http://localhost:8080"
+
+        elif req.headers.get(hdrs.ORIGIN) == "https://cloud.temporal.io":
+            logger.info("Setting CORS headers for cloud.temporal.io")
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_ORIGIN] = "https://cloud.temporal.io"
+
+        allow_headers = "content-type,x-namespace"
+        if req.scheme.lower() == "https":
+            allow_headers += ",authorization"
+            resp.headers[hdrs.ACCESS_CONTROL_ALLOW_CREDENTIALS] = "true"
+
+        # common
+        resp.headers[hdrs.ACCESS_CONTROL_ALLOW_METHODS] = "POST"
+        resp.headers[hdrs.ACCESS_CONTROL_ALLOW_HEADERS] = allow_headers
+
+        return resp
+
+    def decryption_authorized(email: str, namespace: str) -> bool:
+        credentials = grpc.composite_channel_credentials(grpc.ssl_channel_credentials(
+        ), grpc.access_token_call_credentials(os.environ.get("TEMPORAL_API_KEY")))
+
+        with grpc.secure_channel(temporal_ops_address, credentials) as channel:
+            client = service_pb2_grpc.CloudServiceStub(channel)
+            request = request_response_pb2.GetUsersRequest()
+
+            response = client.GetUsers(request, metadata=(
+                ("temporal-cloud-api-version", os.environ.get("TEMPORAL_OPS_API_VERSION")),))
+
+            authorized = False
+            for user in response.users:
+                if user.spec.email.lower() == email.lower():
+                    if user.spec.access.account_access.role in AUTHORIZED_ACCOUNT_ACCESS_ROLES:
+                        authorized = True
+                    else:
+                        if namespace in user.spec.access.namespace_accesses:
+                            if user.spec.access.namespace_accesses[namespace].permission in AUTHORIZED_NAMESPACE_ACCESS_ROLES:
+                                authorized = True
+
+            return authorized
+
+    def make_handler(fn: str):
+        async def handler(req: web.Request):
+            # Read payloads as JSON
+            assert req.content_type == "application/json"
+            payloads = json_format.Parse(await req.read(), Payloads())
+
+            # Extract the email from the JWT.
+            auth_header = req.headers.get("Authorization")
+            namespace = req.headers.get("x-namespace")
+            _bearer, encoded = auth_header.split(" ")
+            decoded = jwt.decode(encoded, options={"verify_signature": False})
```

Comment:
Are we sure we don't want to verify this signature against the cloud's public key? Ref https://docs.temporal.io/production-deployment/data-encryption#authorization, we probably want to check against https://login.tmprl.cloud/.well-known/jwks.json or it's subject to MITM.